### ISSUE TITLE
ServiceCollection.addSingleton and others missing

### ISSUE THREAD

**Issue by andriwas** [paulsgcross/haxe-injection#22]
ServiceCollection.addSingleton and others missing

**Comment by paulsgcross** [comment]
Hi there,

Try adding `using hx.injection.ServiceExtensions` to the top of your class that is using the service collection.

Since ver 3.0.0, `addSingleton` (and all others) have been moved out of the `ServiceCollection` class and into the aforementioned extensions class to reduce cluttering.

**Comment by andriwas** [comment]
got another error after fixing the error by adding `using hx.injection.ServiceExtensions`;
`...Field getConstructorArgs is declared 'override' but parent field getConstructorArgs is 'abstract' and does not provide any implementation to override`

**Comment by paulsgcross** [comment]
Would it be possible to see a snippet of the base class and any extensions so that I can diagnose the error?

**Comment by andriwas** [comment]
sure,
Debug.hx:
```
package;

import hx.injection.Service;
import hx.injection.ServiceCollection;
import hx.injection.ServiceType;
import tink.CoreApi.Callback;

// this works in v2.1.0, comment this when using version 2.1.0
using hx.injection.ServiceExtensions;
// ------------------------------------------------------------

enum abstract ServiceStatus(Int)
{
	public var Ok = 123;
}

enum abstract ServiceBaseErrCodes(Int)
{
	public var OK = 0;
}

interface IService extends Service
{
	public function Start():ServiceStatus;
	public function Pause():ServiceStatus;
	public function Stop():ServiceStatus;
}

interface IServiceBase<TServiceErrCodeEnum> extends IService
{
	public var TargetState(default, null):ServiceStatus;
	public var CurrentState(default, null):ServiceStatus;
	public var ErrCode(default, null):TServiceErrCodeEnum;
}

interface INodeBase<TServiceErrCodeEnum> extends IServiceBase<TServiceErrCodeEnum> {}

abstract class NodeBase<ServiceErrCodeEnum> implements INodeBase<ServiceErrCodeEnum>
{
	public var Status(default, null):ServiceStatus;
	public var ErrCode(default, null):ServiceErrCodeEnum;

	public abstract function Start():ServiceStatus;

	public abstract function Pause():ServiceStatus;

	public abstract function Stop():ServiceStatus;

	public abstract function SendMessage(message:String):Bool;
}

class Node extends NodeBase<ServiceBaseErrCodes>
{
	public function new() {}

	public function Start():ServiceStatus
	{
		return ServiceStatus.Ok;
	}

	public function Pause():ServiceStatus
	{
		return ServiceStatus.Ok;
	}

	public function Stop():ServiceStatus
	{
		return ServiceStatus.Ok;
	}

	public function SendMessage(message:String):Bool
	{
		return true;
	}
}

class Debug
{
	public static function main()
	{
		var collection:ServiceCollection = new ServiceCollection();
		collection.addSingleton(IService, Node);
		collection.createProvider().getService(IService).Start();
	}
}
```

build.hxml
```
-lib haxe-injection
--hl Debug.hl
--main Debug

```

**Comment by andriwas** [comment]
sure,
Debug.hx:
```
package;

import hx.injection.Service;
import hx.injection.ServiceCollection;
import hx.injection.ServiceType;

// comment this when using version 2.1.0
using hx.injection.ServiceExtensions;
// ------------------------------------------------------------

enum abstract ServiceStatus(Int)
{
	public var Ok = 123;
}

enum abstract ServiceBaseErrCodes(Int)
{
	public var OK = 0;
}

interface IService extends Service
{
	public function Start():ServiceStatus;
	public function Pause():ServiceStatus;
	public function Stop():ServiceStatus;
}

interface IServiceBase<TServiceErrCodeEnum> extends IService
{
	public var TargetState(default, null):ServiceStatus;
	public var CurrentState(default, null):ServiceStatus;
	public var ErrCode(default, null):TServiceErrCodeEnum;
}

interface INodeBase<TServiceErrCodeEnum> extends IServiceBase<TServiceErrCodeEnum> {}

abstract class NodeBase<ServiceErrCodeEnum> implements INodeBase<ServiceErrCodeEnum>
{
	public var Status(default, null):ServiceStatus;
	public var ErrCode(default, null):ServiceErrCodeEnum;

	public abstract function Start():ServiceStatus;

	public abstract function Pause():ServiceStatus;

	public abstract function Stop():ServiceStatus;

	public abstract function SendMessage(message:String):Bool;
}

class Node extends NodeBase<ServiceBaseErrCodes>
{
	public function new() {}

	public function Start():ServiceStatus
	{
		return ServiceStatus.Ok;
	}

	public function Pause():ServiceStatus
	{
		return ServiceStatus.Ok;
	}

	public function Stop():ServiceStatus
	{
		return ServiceStatus.Ok;
	}

	public function SendMessage(message:String):Bool
	{
		return true;
	}
}

class Debug
{
	public static function main()
	{
		var collection:ServiceCollection = new ServiceCollection();
		collection.addSingleton(IService, Node);
		collection.createProvider().getService(IService).Start();
	}
}
```

build.hxml
```
-lib haxe-injection
--hl Debug.hl
--main Debug

```

**Comment by paulsgcross** [comment]
This looks like a bug that has been introduced since changing how the macro resolves extensions. I will look into this and push it as a hotfix :)

**Comment by andriwas** [comment]
...

**Comment by paulsgcross** [comment]
Apologies -- I have not forgotten, I have just been laiden with illness and work. I'd appreciate not being bumped with a passive aggressive "..." but instead just ask me if there's an update still coming in plain english.

That being said, I have found a work-around for your problem. Simply add a constructor to your 'NodeBase' class. Then add 'super()' call to your 'Node' class that extends the abstract class like so:
```
class Node extends NodeBase<ServiceBaseErrCodes>
{
	public function new() {
             super();
        }
...
}
```
As the library should support your original way of doing things, I am working on a fix as we speak.

**Comment by paulsgcross** [comment]
Okay, I've fixed the issue. Sorry for the delay. You should now be able to use the library as intended in your example. Patched version is here: https://lib.haxe.org/p/haxe-injection/3.0.1/
